dask/array/random.py

```diff
@@ -17,7 +17,603 @@
+            if isinstance(ar, Array):
+                return broadcast_to(ar, shape).rechunk(chunks)
+            if isinstance(ar, np.ndarray):
+                return np.ascontiguousarray(np.broadcast_to(ar, shape))
+
+        # Broadcast all arguments, get tiny versions as well
+        # Start adding the relevant bits to the graph
+        dsk = {}
+        lookup = {}
+        small_args = []
+        dependencies = []
+        for i, ar in enumerate(args):
+            if isinstance(ar, (np.ndarray, Array)):
+                res = _broadcast_any(ar, size, chunks)
+                if isinstance(res, Array):
+                    dependencies.append(res)
+                    lookup[i] = res.name
+                elif isinstance(res, np.ndarray):
+                    name = f"array-{tokenize(res)}"
+                    lookup[i] = name
+                    dsk[name] = res
+                small_args.append(ar[tuple(0 for _ in ar.shape)])
+            else:
+                small_args.append(ar)
+
+        small_kwargs = {}
+        for key, ar in kwargs.items():
+            if isinstance(ar, (np.ndarray, Array)):
+                res = _broadcast_any(ar, size, chunks)
+                if isinstance(res, Array):
+                    dependencies.append(res)
+                    lookup[key] = res.name
+                elif isinstance(res, np.ndarray):
+                    name = f"array-{tokenize(res)}"
+                    lookup[key] = name
+                    dsk[name] = res
+                small_kwargs[key] = ar[tuple(0 for _ in ar.shape)]
+            else:
+                small_kwargs[key] = ar
+
+        sizes = list(product(*chunks))
+        seeds = random_rng_data(len(sizes), self._bit_generator_state)
+        token = tokenize(seeds, size, chunks, args, kwargs)
+        name = f"{funcname}-{token}"
+
+        keys = product(
+            [name], *([range(len(bd)) for bd in chunks] + [[0]] * len(extra_chunks))
+        )
+        blocks = product(*[range(len(bd)) for bd in chunks])
+
+        vals = []
+        for seed, size, slc, block in zip(seeds, sizes, slices, blocks):
+            arg = []
+            for i, ar in enumerate(args):
+                if i not in lookup:
+                    arg.append(ar)
+                else:
+                    if isinstance(ar, Array):
+                        arg.append((lookup[i],) + block)
+                    else:  # np.ndarray
+                        arg.append((getitem, lookup[i], slc))
+            kwrg = {}
+            for k, ar in kwargs.items():
+                if k not in lookup:
+                    kwrg[k] = ar
+                else:
+                    if isinstance(ar, Array):
+                        kwrg[k] = (lookup[k],) + block
+                    else:  # np.ndarray
+                        kwrg[k] = (getitem, lookup[k], slc)
+            vals.append(
+                (_apply_random_func, self._generator, funcname, seed, size, arg, kwrg)
+            )
```

Comment:
Given this is a brand new function, what about making this HLG aware to start with? Namely instead of generating the chunk-wise graph, we provide a way to defer generation of that graph until this reaches the Scheduler

Reply:
again same as above. prefer to keep as it is if there arent any strong objections